docs/01-tooling/01-automatic-file-bundling.md

```diff
@@ -1,6 +1,188 @@
+that:
+
+* We would use a node based file bundler ([webpack][]), in order to:
+  * automatically bundle our JS sources with a single source of truth (the
+    source code itself)
+  * use a standard module system (we started with common.js and afterwards
+    migrated to ES modules when finalized)
+  * validate the dependency graph and asset building before going to production
+    (webpack statically analyzes the asset tree and fails on build)
+* We would introduce a build step to compile sources into production worthy
+  assets
+* We would use ResourceLoader as the way to **serve** the JS bundle(s) and to
+  specify other dependencies (images, i18n messages, styles)
+
+You can read some more details in the architecture design record [4. Use
+webpack][use-webpack] that we wrote when we discussed this in the team.
+
+#### Why webpack and not `<my-favorite-tool>`?
+
+We looked at the ecosystem of bundlers at the time, and this is a summary of
+our evaluation:
+
+* Browserify: Great tool, shrinking community and support in favor of others
+* Rollup: Great tool, specialized in bundling code for producing libraries, not
+  applications
+* Webpack: Great tool, community, maintainers, financial support, mindshare and
+  ecosystem of tools
+
+We are not married to, or using any webpack specific features, so we can migrate
+from this tool in the future if it becomes a problem.
+
 ### Results
+
+* We introduced `webpack` ([config][webpack-config]) and [build scripts][]
+* The JS files are located in [src/][src] and use EcmaScript modules
+  ([example][es-modules-example])
+* The production assets get built into [/resources/dist][dist] to be served by
+  ResourceLoader
+* We added a CI step in the `npm-test` job that checks built assets have been
+  commited and up to date ([check-built-assets][check-built-assets])
+* Added a precommit hook that automatically runs the build step and adds the
+  built assets ([precommit][precommit])
+
+These are some of the benefits we have seen after the introduction of these
+changes:
+
+* ResourceModules configuration reduced for JS files, file order is automatic
+  now, source code is the single truth for dependency order
+  * [Before][scripts-order], and [after][scripts-after] (no manual ordering of
+    files)
+* Decreased cost and burden for splitting modules, reusing code, adding new code
+  since file order is automatically sorted out
+* Source maps in development that point to the original modules in `src/` thanks
+  to webpack
+* Anecdotally, faster ResourceLoader performance by offloading file parsing and
+  bundling to the node CLI tool watcher (`npm start`)
+* We bundle `redux` and `redux-thunk` from npm, based on the [pinned
+  versions][redux-deps] from `package.json`
+  * `npm outdated` will show if there are new versions of the libraries
+  * `npm update redux` will update the library, and be bundled in our sources
+    when we `npm run build`
+
```

Comment:
* Unit tests that run in an order of magnitude (sometimes two orders of magnitude) faster than the MediaWiki-run ones, which gives engineers a faster feedback loop.

Reply:
I'll expand a lot more on 03-faster-and-isolated-unit-tests but you're totally right, I've adapted your comment to:

```
* By using standard module systems, it enables us to use our sources in node.js,
  which unlocks the possibility to run unit tests for the frontend code in
  node.js for faster test runs and feedback loop for developers
```